src/main/java/com/thevoxelbox/voxelsniper/brush/PerformBrush.java

```diff
@@ -105,54 +103,72 @@ protected boolean perform(SnipeData v, int x, int y, int z) {
             case TYPE:
-                if (current.getType() != v.getReplaceIdState().getType()) {
+                if (!sameBlockType(current, v.getReplaceState())) {
                     return false;
                 }
                 break;
-            case STATE:
-                // @Todo filter by key and value
+            case TRAITS:
+                if (!BlockHelper.hasTraits(current, v.getReplaceInkTraits())) {
+                    return false;
+                }
                 break;
             case COMBO:
-                if (current != v.getReplaceIdState()) {
+                if (!sameBlockType(current, v.getReplaceState()) ||
+                    !BlockHelper.hasTraits(current, v.getReplaceInkTraits())) {
                     return false;
                 }
                 break;
             case NONE:
             default:
                 break;
-            }
         }
-        switch (this.place) {
-        case TYPE:
-            setBlockType(x, y, z, v.getVoxelIdState().getType(), this.physics ? BlockChangeFlags.ALL : BlockChangeFlags.NONE);
-            break;
-        case STATE:
-            BlockState current = this.world.getBlock(x, y, z);
-            @SuppressWarnings({"unchecked", "rawtypes"})
-            Optional<BlockState> place = current.with((Key) v.getVoxelInkKey(), v.getVoxelInkValue());
-            if (!place.isPresent()) {
-                return false;
-            }
-            setBlockState(x, y, z, place.get(), this.physics ? BlockChangeFlags.ALL : BlockChangeFlags.NONE);
-            break;
-        case COMBO:
-            setBlockState(x, y, z, v.getVoxelIdState(), this.physics ? BlockChangeFlags.ALL : BlockChangeFlags.NONE);
-            break;
-        case NONE:
-        default:
-            throw new IllegalStateException("Unsupported place type " + this.place.name());
+
+        BlockChangeFlag physicsFlags = usePhysics ? BlockChangeFlags.ALL : BlockChangeFlags.NONE;
+        switch (placeMethod) {
+            case TYPE:
+                setBlockType(x, y, z, v.getVoxelState().getType(), physicsFlags);
+                break;
+            case TRAITS:
+                BlockState place = BlockHelper.addTraits(current, v.getVoxelInkTraits());
+                setBlockState(x, y, z, place, physicsFlags);
+                break;
+            case COMBO:
+                setBlockState(x, y, z, v.getVoxelState(), physicsFlags);
+                break;
+            case NONE:
+            default:
+                throw new IllegalStateException("Unsupported place type " + placeMethod.name());
         }
         return true;
     }
 
-    public static enum PerformerType {
+    private boolean sameBlockType(BlockState a, BlockState b) {
+        return a.getType().equals(b.getType());
+    }
+
+    public enum PerformerType {
         TYPE,
-        STATE,
+        TRAITS,
         COMBO,
         NONE;
+
+        public String toString() {
+            switch (this) {
```

Comment:
This should just be a field on the enum.

Reply:
What do you mean by that? The toString method is already part of the enum class.